engine/runtime-compiler/src/test/java/org/enso/compiler/test/pass/MiniPassTraverserTest.java

```diff
@@ -0,0 +1,108 @@
+package org.enso.compiler.test.pass;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+import java.util.List;
+import org.enso.compiler.pass.MiniIRPass;
+import org.junit.Test;
+
+public class MiniPassTraverserTest {
+  @Test
+  public void traversesOneExpression() {
+    var expr = new MockExpression(false);
+    var miniPass = MockMiniPass.builder().build();
+    MiniIRPass.compile(MockExpression.class, expr, miniPass);
+    assertThat(
+        "Prepare is called only for trees with depth > 1", expr.isPreparedByAny(), is(false));
+    assertThat(expr.isTransformedByAny(), is(true));
+  }
+
+  @Test
+  public void traversesExpressionWithOneChild() {
+    var parentExpr = new MockExpression(false);
+    var childExpr = new MockExpression(true);
+    parentExpr.addChild(childExpr);
+    var miniPass = MockMiniPass.builder().build();
+    MiniIRPass.compile(MockExpression.class, parentExpr, miniPass);
+    assertThat(
+        "Prepare must be called on a child expression", childExpr.isPreparedByAny(), is(true));
+    assertThat(childExpr.isTransformedByAny(), is(true));
+    assertThat(parentExpr.isTransformedByAny(), is(true));
+  }
+
+  @Test
+  public void traversesExpressionWithManyChildren() {
+    var parentExpr = new MockExpression(false);
+    var children = List.of(new MockExpression(true), new MockExpression(true));
+    children.forEach(parentExpr::addChild);
+    var miniPass = MockMiniPass.builder().build();
+    MiniIRPass.compile(MockExpression.class, parentExpr, miniPass);
+    for (var ch : children) {
+      assertThat("Prepare must be called on a child expression", ch.isPreparedByAny(), is(true));
+      assertThat(ch.isTransformedByAny(), is(true));
+    }
+    assertThat(parentExpr.isTransformedByAny(), is(true));
+  }
+
+  @Test
+  public void stopTraversingWhenPrepareReturnsNull() {
+    var e1 = new MockExpression(false);
+    var e2 = new MockExpression(true);
+    var e3 = new MockExpression(true);
+    e1.addChild(e2);
+    e2.addChild(e3);
+    // Should stop traversing when e3 is encountered.
+    // Should only process e1 and e2, not e3
+    var miniPass = MockMiniPass.builder().stopExpr(e3).build();
+    MiniIRPass.compile(MockExpression.class, e1, miniPass);
+    assertThat("e3 should not be processed", e3.isPreparedByAny(), is(false));
+    assertThat("e3 should not be processed", e3.isTransformedByAny(), is(false));
```

Comment:
I am missing a test that would check a `ChainedMiniPass` is able to skip some elements in the tree - e.g. that there can be `!element.isTransformedByAny()` in an `IR` traversed by a chained mini pass. Reading the current code I assume such a test should fail.

Great that the behavior of `MiniPassTraverser` is tested at all!

Reply:
Added such a test in 023d1aa4165a805b7f5273209601f98435b9d4cc